src/webview-js/toc-editor/toc-editor.jsx

```diff
@@ -0,0 +1,155 @@
+import { h, Fragment, render } from 'preact';
+import { useEffect, useState } from 'preact/hooks';
+import 'react-sortable-tree/style.css';
+import SortableTree from 'react-sortable-tree';
+
+const vscode = acquireVsCodeApi();
+const nodeType = 'toc-element'
+
+const removeExpanded = (obj) => {
+  if (typeof obj !== 'object') {
+    return obj
+  }
+  if (obj instanceof Array) {
+    const copy = []
+    for (const subobj of obj) {
+      copy.push(removeExpanded(subobj))
+    }
+    return copy
+  }
+  const { expanded, ...copy } = obj
+  copy.children = removeExpanded(copy.children)
+  return copy
+}
+
+const ContentTree = (props) => {
+  const modifiesStateName = props.modifiesStateName
+  const [data, setData] = useState(props.data);
+
+  const handleChange = (newChildren) => {
+    const { treesData, selectionIndices } = vscode.getState()
+
+    const newData = { ...data, ...{children: newChildren} }
+    const oldStructure = JSON.stringify(removeExpanded(data.children))
+    const newStructure = JSON.stringify(removeExpanded(newChildren))
+
+    if (data.children.length - newChildren.length > 3) {
+      // There's a bug that deletes the whole tree except for one element.
+      // Prevent this by not allowing high magnitude deletions
+      return
+    }
+
+    treesData[modifiesStateName][props.index].children = newChildren
+    vscode.setState({ treesData, selectionIndices })
+    setData(newData)
+
+    if (oldStructure !== newStructure) {
+      vscode.postMessage({treeData: newData})
+    }
+  }
+
+  const getNodeProps = ({ node }) => {
+    const typeToColor = {
+      'subcollection': 'green',
+      'module': 'purple'
+    }
+    return {
+      style: {
+        boxShadow: `0 0 0 4px ${typeToColor[node.type]}`,
+      },
+      title: node.title + (node.moduleid ? ` (${node.moduleid})` : '')
+    }
+  }
+
+  const canDrop = ({ nextParent}) => {
+    if (nextParent && nextParent.type === 'module') {
+      return false
+    }
+    return true
+  }
+  
+  return (
+    <div style={{ height: '100%' }}>
+      <SortableTree
+        treeData={data.children}
+        onChange={props.editable ? handleChange : () => {}} // non-editable treeData will never change
+        generateNodeProps={getNodeProps}
+        canDrop={props.editable ? canDrop : () => true} // Dropping item into non-editable trees will destroy the item
```

Comment:
Is that expected behavior?

Reply:
I think so. Dragging from non-editable trees is how you add elements to a collection. Seems like to remove it, you'd just "put it back".  It's also the only way to delete elements right now.